### PR TITLE
Using attributes to enable mirroring for an operator

### PR DESCRIPTION
**Pull request by pluskid**
The user can now put `{'force_mirroring': 'true'}` as attributes to force mirroring of a particular operator.

@tqchen I got the following observations, did you observe similar thing when coding up the mirroring feature?

- Some operators (e.g. `BatchNorm`) will crash when mirroring is turned on, with some error like `src/symbol/graph_executor.cc:606: Check failed: (info.type) != (kNotInitialized)`.
- The memory consumption is a bit mysterious to measure, for the cifar10 example. Although I'm only using pycuda to query the total gpu free memory, which is a very rough estimation, I could see that running the exactly same network training (without touching any mirroring option) could actually lead to quite different reporting of memory consumption.

### CODE REVIEW

File: src/symbol/static_graph.h [hunk 1]
```diff
@@ -143,6 +144,21 @@ class StaticGraph {
       addto_index = std::move(another.addto_index);
       return *this;
     }
+
+    template<typename ValueType>
+    inline ValueType get_attr(const std::string& key, ValueType default_value) const {
```
Comment (tqchen) [inline src/symbol/static_graph.h]: Code Style

Comment (tqchen) [inline src/symbol/static_graph.h]: GetAttr

**Comment by tqchen** [comment]
The code LGTM. @pluskid I would consider the crash as a potential BUG. if you can get a reproducible script say on CIFAR or invariant of CIFAR structure, please send a gist and I will look into it

File: src/symbol/static_graph.h (hunk 1 above)
Comment (pluskid) [inline src/symbol/static_graph.h]: This function is for internal use (not C API). I'm following other functions here (see e.g. `is_forward`) right below it.

**Comment by pluskid** [comment]
@tqchen Sure, here is a simple example. On this branch, if you change the line 10 of `symbol_inception-bn-28-small.py` to
```
conv = mx.symbol.Convolution(data=data, num_filter=num_filter, kernel=kernel, stride=stride, pad=pad, attr=mirror_attr)
```
basically, add the force mirroring attr to the convolution layer. Then run `train_cifar10_mirroring.py`, you will get
```
mxnet/mxnet/dmlc-core/include/dmlc/logging.h:241: [18:28:02] src/symbol/graph_executor.cc:606: Check failed: (info.type) != (kNotInitialized)
Traceback (most recent call last):
  File "train_cifar10_mirroring.py", line 114, in <module>
    train_model.fit(args, net_mirrored, get_iterator, batch_end_callback=report_gpu_memory())
  File "mxnet/mxnet/example/image-classification/train_model.py", line 100, in fit
    epoch_end_callback = checkpoint)
  File "mxnet/mxnet/python/mxnet/model.py", line 770, in fit
    sym_gen=self.sym_gen)
  File "mxnet/mxnet/python/mxnet/model.py", line 191, in _train_multi_device
    logger=logger)
  File "mxnet/mxnet/python/mxnet/executor_manager.py", line 277, in __init__
    self.slices, train_data)
  File "mxnet/mxnet/python/mxnet/executor_manager.py", line 190, in __init__
    shared_data_arrays=self.shared_data_arrays[i])
  File "mxnet/mxnet/python/mxnet/executor_manager.py", line 143, in _bind_exec
    grad_req='write' if need_grad else 'null', shared_exec=base_exec)
  File "mxnet/mxnet/python/mxnet/symbol.py", line 774, in bind
    ctypes.byref(handle)))
  File "mxnet/mxnet/python/mxnet/base.py", line 77, in check_call
    raise MXNetError(py_str(_LIB.MXGetLastError()))
mxnet.base.MXNetError: [18:28:02] src/symbol/graph_executor.cc:606: Check failed: (info.type) != (kNotInitialized)
```

**Comment by tqchen** [comment]
I am still traveling and won't be able to rerun this till Sat. I will think about what might be the cause for now. In the meanwhile, let us aim to merge this in when the issues are fixed. @pluskid

**Pull request merged**